File: composeApp/src/androidMain/kotlin/uk/co/oliverdelange/location_alarm/screens/debug/Logs.kt
package uk.co.oliverdelange.location_alarm.screens.debug

import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import uk.co.oliverdelange.location_alarm.R
import uk.co.oliverdelange.location_alarm.ui.AutoScrollToBottom
import uk.co.oliverdelange.locationalarm.model.ui.debug.LogColor
import uk.co.oliverdelange.locationalarm.model.ui.debug.LogEntryUiModel
import uk.co.oliverdelange.locationalarm.model.ui.debug.LogUiState

@Composable
fun Logs(state: LogUiState) {
    AutoScrollToBottom(state.logs.size) { scrollState ->
        LazyColumn(
            state = scrollState,
            modifier = Modifier
                .padding(8.dp)
        ) {
            items(state.logs) {
                val string = buildAnnotatedString {
                    withStyle(SpanStyle(fontWeight = FontWeight.Bold)) {
                        append(it.date + "\t")
                    }
                    withStyle(SpanStyle(color = getLogColor(it))) {
                        append(it.message)
                    }
                }

                Text(string, fontSize = 10.sp)
            }
        }
    }
}

@Composable
private fun getLogColor(it: LogEntryUiModel) = when (it.color) {
    LogColor.Red -> colorResource(R.color.red)
    LogColor.Orange -> colorResource(R.color.orange)
    LogColor.Green -> colorResource(R.color.green)
    LogColor.Blue -> colorResource(R.color.blue)
    LogColor.Grey -> colorResource(R.color.grey)
}

File: iosApp/iosApp/screens/debug/Logs.swift
import SwiftUI
import KMPObservableViewModelSwiftUI
import Shared

struct Logs: View {
    var logUiState: LogUiState
    @State private var atBottom: Bool = true
    
    var body: some View {
        ScrollViewReader { scrollView in
            ScrollView{
                LazyVStack(alignment: .leading) {
                    ForEach(logUiState.logs, id: \.self) { log in
                        let logColor = getLogColor(log.color)
                        ZStack {
                            Text(log.date + "\t")
                                .font(.caption2)
                                .fontWeight(.bold)
                            + Text(log.message)
                                .font(.caption)
                                .foregroundColor(logColor)
                        }
                        .id(log)
                    }
                    Color.clear
                        .frame(width: 0, height: 0, alignment: .bottom)
                        .onAppear { atBottom = true }
                        .onDisappear { atBottom = false }
                }.padding(8)
                    .onChange(of: logUiState.logs) { _ in
                        // Automatically scroll to the bottom when items change
                        if let lastItem = logUiState.logs.last {
                            if (atBottom) { scrollView.scrollTo(lastItem, anchor: .bottom) }
                        }
                    }
            }
        }
    }
    
    func getLogColor(_ logColor: LogColor) -> Color {
        switch(logColor){
        case .red:
            return Color(.red)
        case .orange:
            return Color(.orange)
        case .green:
            return Color(.green)
        case .blue:
            return Color(.blue)
        case .grey:
            return Color(.grey)
        default:
            return Color(.black)
        }
    }
}

#Preview {
    Logs(logUiState: LogUiState())
}



